numpy/core/src/umath/loops.c.src

```diff
@@ -449,21 +449,13 @@ NPY_NO_EXPORT NPY_GCC_OPT_3 int
                       char **args, npy_intp const *dimensions, npy_intp const *steps,
                       void *NPY_UNUSED(func))
 {
-    char *ip1 = args[0];
-    char *indxp = args[1];
-    char *value = args[2];
-    npy_intp is1 = steps[0], isindex = steps[1], isb = steps[2];
-    npy_intp shape = steps[3];
+    @type@ **indexed = (@type@ **)args[0];
```

Comment:
I see how nice it is to simplify the loops, but of course it is not surprising that this leads to a delay, since there now is the intermediate step of creating the index array.

Reply:
Maybe there is a different way to refactor the indexing loops to reduce the repetitions with a macro?

Reply:
Right, you can make the adjustment a macro.  I am not quite sure what you mean by this below (in terms of preference):
> I kind of like the locality of the fast indexing loop

I like that this change separates the concerns a bit.  That gives us easy handling of higher dimensions and to me seems slightly nicer as public API?
The advantage of the old loop is that it is close to what the user passes.

The downside is the slight slowdown compared to current trivial loops (maybe ~10% for advanced indexing, ~25% for `np.add.at`).

Reply:
Like @mattip, I felt that the repetition in these loops of calculating `indx` (or even `indexed`) could be a macro, which would result in a similarly nice "look". And the locality is good because it means the index calculations are all done in registers, not stored in chunks in an array.

But of course what you have here is really nice too, as it allows more complicated index arrays...